### PR TITLE
Do not use metadata deletes during Iceberg writes

### PR DESCRIPTION
**Pull request by alexjo2144**
## Description

This is an alternative to https://github.com/trinodb/trino/pull/18412 but with more minimal changes.

## Additional context and related issues

Fixes: https://github.com/trinodb/trino/issues/18393

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
some tests have failed

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java [hunk 1]
```diff
@@ -154,8 +154,8 @@ public void testAnalyzeWithSchemaEvolution()
                         VALUES
                           ('nationkey', null, 25, 0, null, '0', '24'),
                           ('regionkey', null, 5, 0, null, '0', '4'),
-                          ('name', 1314.0, 25, 0, null, null, null),
-                          ('info', 4417.0, null, 0, null, null, null),
+                          ('name', 1907.9999999999998, 25, 0, null, null, null),
```
Comment (ksobolew) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java:157]: That does not look like reliably repeatable result

Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java:157]: Yeah, it's also different when I run the test locally.

**Comment by alexjo2144** [comment]
@findepi last build was green, just squashed

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java (hunk 1 above)
Comment (findepi) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java:157]: does it mean the test doesn't pass locally, or is flaky?

Comment (findepi) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java:157]: BTW double results are asserted with rounding. you should probably have 1908.0 instead of 1907.9999999999998
(locally the test passed for me with both values; i would use the shorter one)

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java [hunk 2]
```diff
@@ -165,9 +165,9 @@ public void testAnalyzeWithSchemaEvolution()
                         VALUES
                           ('nationkey', null, 25, 0, null, '0', '24'),
                           ('regionkey', null, 5, 0, null, '0', '4'),
-                          ('name', 1314.0, 25, 0, null, null, null),
-                          ('info', 4417.0, 25, 0, null, null, null),
-                          (null, null, null, null, 25, null, null)""");
+                          ('name', 1907.9999999999998, 25, 0, null, null, null),
+                          ('info', 4417.0, 25, 0.1, null, null, null),
+                          (null, null, null, null, 50, null, null)""");
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java:170]: add a comment why 50

i am surprised `testAnalyzeWithSchemaEvolution` is the only method that shows a user-visible behavioral regression (stats inaccuracy after whole file deleted), especially that it has nothing to with schema evolution.  Please add simpler dedicated test that shows that.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java [hunk 3]
```diff
@@ -339,6 +350,98 @@ public void testOptimizingPartitionsOfV2TableWithGlobalEqualityDeleteFile()
                                 .toArray(String[]::new));
     }
 
+    @Test
+    public void runOptimizeDuringWriteOperations()
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java:354]: test methods typically start with a "test"

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java [hunk 4]
```diff
@@ -113,6 +123,7 @@ protected QueryRunner createQueryRunner()
         return IcebergQueryRunner.builder()
                 .setInitialTables(NATION)
                 .setMetastoreDirectory(metastoreDir)
+                .addAdditionalCatalog(new BlackHolePlugin(), "blackhole", "blackhole")
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java:126]: `IcebergQueryRunner.addAdditionalCatalog` is redundant, you can achieve the same with
```
queryRunner.installPlugin(new BlackHolePlugin());
queryRunner.createCatalog(...);
```

(currently requires some boilerplate to do robustly
```java
DistributedQueryRunner queryRunner = IcebergQueryRunner.builder()
        .setInitialTables(NATION)
        .setMetastoreDirectory(metastoreDir)
        .build();
try {
    queryRunner.installPlugin(new BlackHolePlugin());
    queryRunner.createCatalog("blackhole", "blackhole");
} catch (Exception e) {
    Closables.closeAllSuppress(e, queryRunner::close);
    throw e;
}
return queryRunner;
```
)

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/IcebergQueryRunner.java [hunk 5]
```diff
@@ -123,6 +126,12 @@ public Builder setSchemaInitializer(SchemaInitializer schemaInitializer)
             return self();
         }
 
+        public Builder addAdditionalCatalog(Plugin connectorPlugin, String catalogName, String connectorName)
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/IcebergQueryRunner.java:129]: if we wanted to have this, this would go into parent DQR, not specific to iceberg.

for now let's remove it, see comment in TIV2.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorSmokeTest.java [hunk 6]
```diff
@@ -124,28 +128,38 @@ public void testDeleteRowsConcurrently()
         int threads = 4;
         CyclicBarrier barrier = new CyclicBarrier(threads);
         ExecutorService executor = newFixedThreadPool(threads);
+        List<String> rows = ImmutableList.of("(1, 0, 0, 0)", "(0, 1, 0, 0)", "(0, 0, 1, 0)", "(0, 0, 0, 1)");
+
+        String[] expectedErrors = new String[]{"Failed to commit Iceberg update to table:", "Failed to replace table due to concurrent updates:"};
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorSmokeTest.java:133]: `verifyConcurrentUpdateFailurePermissible` allows  `"Failed to commit Iceberg update to table:"` only. is the latter possible as well?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorSmokeTest.java [hunk 7]
```diff
@@ -124,28 +128,38 @@ public void testDeleteRowsConcurrently()
...
         try (TestTable table = new TestTable(
                 getQueryRunner()::execute,
                 "test_concurrent_delete",
                 "(col0 INTEGER, col1 INTEGER, col2 INTEGER, col3 INTEGER)")) {
             String tableName = table.getName();
-            assertUpdate("INSERT INTO " + tableName + " VALUES (0, 0, 0, 0)", 1);
-            assertUpdate("INSERT INTO " + tableName + " VALUES (1, 0, 0, 0)", 1);
-            assertUpdate("INSERT INTO " + tableName + " VALUES (0, 1, 0, 0)", 1);
-            assertUpdate("INSERT INTO " + tableName + " VALUES (0, 0, 1, 0)", 1);
-            assertUpdate("INSERT INTO " + tableName + " VALUES (0, 0, 0, 1)", 1);
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorSmokeTest.java:136]: The test created 5 files, now it creates ~1 file with 4 rows.
If this doesn't matter for the test purpose, let's do this change as a separate commit.
if this matters and should always be so (~1 file), let's do this as a separate (prep) commit.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 8]
```diff
@@ -1991,6 +1992,7 @@ public void testPartitionPredicatePushdownWithHistoricalPartitionSpecs()
 
         // The old partition scheme is no longer used so pushdown using the hour transform is allowed
         assertUpdate("DELETE FROM " + tableName + " WHERE year(d) = 1969", 3);
+        assertUpdate("ALTER TABLE " + tableName + " EXECUTE optimize");
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:1995]: why change this test?
(separate commit?)

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 9]
```diff
@@ -1378,6 +1378,7 @@ public void testUpdateWithSortOrder()
                 "test_sorted_update",
                 "WITH (sorted_by = ARRAY['comment']) AS TABLE tpch.tiny.lineitem WITH NO DATA")) {
             assertUpdate(
+                    withSmallRowGroups,
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:1381]: why change this test?
(separate commit?)

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 10]
```diff
@@ -2142,132 +2135,76 @@ private void finishWrite(ConnectorSession session, IcebergTableHandle table, Col
...
+            commit(rowDelta, session);
             transaction.commitTransaction();
         }
         catch (ValidationException e) {
             throw new TrinoException(ICEBERG_COMMIT_ERROR, "Failed to commit Iceberg update to table: " + table.getSchemaTableName(), e);
         }
-        transaction = null;
-    }
-
-    private static boolean fileIsFullyDeleted(List<CommitTaskData> positionDeletes)
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:2263]: 😢  i will miss you

**Review by findepi — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java (hunk 8 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:1995]: The test relied on the behavior that files were fully removed from the manifest during the delete operation, but now they are not until optimize is run.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java (hunk 1 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java:157]: > does it mean the test doesn't pass locally, or is flaky?

Consistently fails. We chatted about this test on Slack the other day

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorSmokeTest.java (hunk 7 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorSmokeTest.java:136]: It used to matter but doesn't any more. One line per file guaranteed that we always used metadata deletes in this test but now we no longer use those.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java (hunk 10 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:2263]: Maybe it will rise from the ashes again...

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java (hunk 1 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergStatistics.java:157]: Is it a double comparison though? I thought assertions on `SHOW STATS` are all varchars
